File: SERVER/WEB_API/src/main/java/isel/ps/employbox/controllers/ChatController.java
package isel.ps.employbox.controllers;

import isel.ps.employbox.exceptions.BadRequestException;
import isel.ps.employbox.model.ModelBinder;
import isel.ps.employbox.model.binder.MessageBinder;
import isel.ps.employbox.model.input.InChat;
import isel.ps.employbox.model.input.InMessage;
import isel.ps.employbox.model.output.OutChat;
import isel.ps.employbox.model.output.OutMessage;
import isel.ps.employbox.services.APIService;
import isel.ps.employbox.services.ChatService;
import isel.ps.employbox.model.entities.Chat;
import org.springframework.web.bind.annotation.*;

import java.util.List;

import static isel.ps.employbox.ErrorMessages.BAD_REQUEST_IDS_MISMATCH;

@RestController
@RequestMapping("/accounts/{id}/chats")
public class ChatController {
    private final ModelBinder<Chat, OutChat, InChat, Long> chatBinder;
    private final ChatService chatService;
    private final MessageBinder messageBinder;
    private final APIService apiService;

    public ChatController(ModelBinder<Chat, OutChat, InChat, Long> chatBinder, ChatService chatService, MessageBinder messageBinder, APIService apiService) {
        this.chatBinder = chatBinder;
        this.chatService = chatService;
        this.messageBinder = messageBinder;
        this.apiService = apiService;
    }

    @GetMapping
    public List<OutChat> getChats (@PathVariable long id){
        return chatBinder.bindOutput(chatService.getAccountChats(id));
    }

    @GetMapping("/{cid}/messages")
    public List<OutMessage> getChatsMessages (@PathVariable long id, @PathVariable long cid, @RequestParam String page) {
        return messageBinder.bindOutput(chatService.getAccountChatsMessages(id,cid, page));
    }

    @PostMapping
    public void createChat(@PathVariable long id, @RequestHeader("apiKey") String apiKey, @RequestBody InChat inChat) {
        if(id != inChat.getAccountIdFirst()) throw new BadRequestException(BAD_REQUEST_IDS_MISMATCH);
        apiService.validateAPIKey(apiKey);
        chatService.createNewChat(
                chatBinder.bindInput(inChat)
        );
    }

    @PostMapping("/{cid}/messages")
    public void createMessage(@PathVariable long id, @PathVariable long cid, @RequestHeader("apiKey") String apiKey, @RequestBody InMessage msg) {
        apiService.validateAPIKey(apiKey);
        chatService.createNewChatMessage(id, cid, messageBinder.bindInput(msg));
    }
}

File: web-api/src/main/java/isel/ps/employbox/controllers/account/ChatController.kt
package isel.ps.employbox.controllers.account

import isel.ps.employbox.ErrorMessages
import isel.ps.employbox.exceptions.BadRequestException
import isel.ps.employbox.model.binders.ChatBinder
import isel.ps.employbox.model.binders.MessageBinder
import isel.ps.employbox.model.entities.Chat
import isel.ps.employbox.model.entities.Message
import isel.ps.employbox.model.input.InChat
import isel.ps.employbox.model.input.InMessage
import isel.ps.employbox.model.output.Collections.HalCollectionPage
import isel.ps.employbox.model.output.OutMessage
import isel.ps.employbox.services.ChatService
import org.springframework.security.core.Authentication
import org.springframework.web.bind.annotation.*
import reactor.core.publisher.Mono

import java.util.concurrent.CompletableFuture

import isel.ps.employbox.ErrorMessages.BAD_REQUEST_IDS_MISMATCH

@RestController
@RequestMapping("/accounts/{id}/chats")
class ChatController(private val chatBinder: ChatBinder, private val chatService: ChatService, private val messageBinder: MessageBinder) {


    @GetMapping("/{cid}/messages")
    fun getChatsMessages(
            @PathVariable id: Long,
            @PathVariable cid: Long,
            @RequestParam(defaultValue = "0") page: Int,
            @RequestParam(defaultValue = "10") pageSize: Int,
            authentication: Authentication
    ): Mono<HalCollectionPage<Message>> {
        val future = chatService.getAccountChatsMessages(id, authentication.name, page, pageSize)
                .thenCompose({ messageCollectionPage -> messageBinder.bindOutput(messageCollectionPage, this.javaClass, id, cid) })
        return Mono.fromFuture<HalCollectionPage<Message>>(future)
    }

    @PostMapping
    fun createChat(@PathVariable id: Long, @RequestBody inChat: InChat, authentication: Authentication): Mono<Chat> {
        if (id != inChat.accountIdFirst) throw BadRequestException(BAD_REQUEST_IDS_MISMATCH)
        return chatService.createNewChat(
                id,
                chatBinder.bindInput(inChat),
                authentication.name
        )
    }

    @PostMapping("/{cid}/messages")
    fun createMessage(@PathVariable id: Long, @PathVariable cid: Long, @RequestBody msg: InMessage, authentication: Authentication): Mono<OutMessage> {
        if (cid != msg.chatId)
            throw BadRequestException(ErrorMessages.BAD_REQUEST_IDS_MISMATCH)
        val future = chatService.createNewChatMessage(id, cid, messageBinder.bindInput(msg), authentication.name)
                .thenCompose(???({ messageBinder.bindOutput() }))
        return Mono.fromFuture<OutMessage>(future)
    }

    @GetMapping("/{cid}/messages/{mid}")
    fun getChatMessage(@PathVariable cid: Long, @PathVariable mid: Long, authentication: Authentication): Mono<OutMessage> {
        val future = chatService.getAccountChatsMessage(cid, mid, authentication.name)
                .thenCompose(???({ messageBinder.bindOutput() }))
        return Mono.fromFuture<OutMessage>(future)
    }
}
